mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java

```diff
@@ -209,14 +209,29 @@ public Receive createReceive() {
 
     private void onAddNewJobArtifactsToCacheRequest(AddNewJobArtifactsToCacheRequest req) {
         try {
-            mantisJobStore.addNewJobArtifactsToCache(req.getClusterID(), req.getArtifacts());
-            jobArtifactsToCache.addAll(req.artifacts);
+            List<ArtifactID> newArtifacts = req.artifacts
+                .stream()
+                .filter(artifactID -> !jobArtifactsToCache.contains(artifactID))
+                .collect(Collectors.toList());
+
+            if (!newArtifacts.isEmpty()) {
+                log.info("Storing and caching new artifacts: {}", newArtifacts);
+
+                mantisJobStore.addNewJobArtifactsToCache(req.getClusterID(), newArtifacts);
```

Comment:
I would probably change the semantics of this as follows:
```
s <- set of artifacts to be cached 

addNewArtifacts(delta):
  s <- s + delta
  store s
  refresh()
```

Reply:
why? let's say the actor crashes before storing the updated `s`, the delta will get lost. Is that what we want?

Reply:
sure, you can change the ordering as you deem fit. But, the main point was about setting the goal state of the cache.